### PR TITLE
[SPARK-31047][SQL] Improve file listing for ViewFileSystem

### PR DESCRIPTION
**Pull request by manuzhang**
### What changes were proposed in this pull request?
Use `listLocatedStatus` when `lnMemoryFileIndex` is listing files from a `ViewFileSystem` which should delegate to that of `DistributedFileSystem`.

### Why are the changes needed?
When `ViewFileSystem` is used to manage several `DistributedFileSystem`, the change will improve performance of file listing, especially when there are many files.

### Does this PR introduce any user-facing change?
No.

### How was this patch tested?
Existing tests.

### CODE REVIEW

**Comment by manuzhang** [comment]
cc @JoshRosen @dongjoon-hyun

**Comment by dongjoon-hyun** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala [hunk 1]
```diff
@@ -313,7 +314,7 @@ object InMemoryFileIndex extends Logging {
         // to retrieve the file status with the file block location. The reason to still fallback
         // to listStatus is because the default implementation would potentially throw a
         // FileNotFoundException which is better handled by doing the lookups manually below.
-        case _: DistributedFileSystem if !ignoreLocality =>
+        case (_: DistributedFileSystem | _: ViewFileSystem) if !ignoreLocality =>
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Hi, @manuzhang .
Which test did you referring in the PR description? Could you add a UT explicitly?
```
## How was this patch tested?

Existing tests.
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala (hunk 1 above)
Comment (manuzhang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: Sorry, I didn't find a UT for `DistributedFileSystem`. Could you suggest on how to test it ?

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: `Mock`?
```
$ git grep 'DistributedFileSystem' | grep Suite
core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:import org.apache.hadoop.hdfs.{DFSInputStream, DistributedFileSystem}
core/src/test/scala/org/apache/spark/deploy/history/FsHistoryProviderSuite.scala:    val dfs = mock(classOf[DistributedFileSystem])
sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala:      "fs.file.impl" -> classOf[MockDistributedFileSystem].getName,
sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala:// This file system is for SPARK-14959 (DistributedFileSystem will throw an exception
sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileSourceStrategySuite.scala:class MockDistributedFileSystem extends RawLocalFileSystem {
```

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: BTW, without a test case, your contribution is not protected from accidental removal. That's the reason why we ask a UT always.

Comment (manuzhang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala]: thanks, UT added

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
Thank you for updating, @manuzhang .

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @manuzhang and @HyukjinKwon .
Merged to master.

**Pull request closed without merging**